File: genai/chatbot.py
import os
from typing import List, Dict, Optional
from rag_pipeline import RAGPipeline, create_educational_knowledge_base

try:
    import openai
    OPENAI_AVAILABLE = True
except ImportError:
    OPENAI_AVAILABLE = False


class StudentPerformanceChatbot:
    def __init__(self, api_key: Optional[str] = None, model: str = 'gpt-3.5-turbo'):
        self.api_key = api_key or os.getenv('OPENAI_API_KEY')
        self.model = model
        self.rag_pipeline = RAGPipeline()
        self.conversation_history = []
        knowledge_base = create_educational_knowledge_base()
        self.rag_pipeline.add_to_knowledge_base(knowledge_base)
        if self.api_key and OPENAI_AVAILABLE:
            openai.api_key = self.api_key
            self.use_openai = True
        else:
            self.use_openai = False

    def index_student_data(self, students_data: List[Dict]):
        self.rag_pipeline.index_student_data(students_data)

    def chat(self, user_message: str, student_id: Optional[str] = None) -> str:
        context = self.rag_pipeline.build_context_for_llm(user_message, student_id)
        if self.use_openai:
            response = self._generate_openai_response(user_message, context)
        else:
            response = self._generate_fallback_response(user_message, student_id)
        self.conversation_history.append({
            'user': user_message,
            'assistant': response,
            'student_id': student_id
        })
        return response

    def _generate_openai_response(self, user_message: str, context: str) -> str:
        try:
            system_prompt = """You are an expert educational advisor and data analyst specializing in student performance. 
Your role is to:
1. Analyze student performance data and provide insights
2. Explain why students might be at risk
3. Suggest evidence-based intervention strategies
4. Answer questions about educational best practices
5. Provide actionable, practical advice for teachers and administrators

Be empathetic, professional, and focus on solutions that help students succeed."""

            messages = [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": f"{context}\n\nQuestion: {user_message}"}
            ]

            response = openai.ChatCompletion.create(
                model=self.model,
                messages=messages,
                temperature=0.7,
                max_tokens=500
            )

            return response.choices[0].message.content.strip()
        except Exception:
            return self._generate_fallback_response(user_message, None)

    def _generate_fallback_response(self, user_message: str, student_id: Optional[str]) -> str:
        message_lower = user_message.lower()
        if student_id:
            student_context = self.rag_pipeline.retrieve_student_context(student_id)
            if 'why' in message_lower and 'risk' in message_lower:
                return self._explain_risk(student_context)
            if 'how' in message_lower and ('improve' in message_lower or 'help' in message_lower):
                return self._suggest_improvements(student_context)
            if 'what' in message_lower and 'intervention' in message_lower:
                return self._suggest_interventions(student_context)

        if 'at risk' in message_lower or 'failing' in message_lower:
            return """At-risk students are those predicted to have academic difficulties based on factors like:
- Low attendance rates
- Declining grades or marks
- Incomplete assignments
- Poor class participation

Early identification allows for timely intervention and support."""

        if 'intervention' in message_lower or 'help' in message_lower:
            return """Effective interventions for at-risk students include:
1. Personalized learning plans tailored to individual needs
2. Regular one-on-one meetings to track progress
3. Peer tutoring and mentorship programs
4. Parent-teacher collaboration and communication
5. Additional academic support and tutoring
6. Study skills and time management training
7. Addressing attendance and engagement issues
8. Setting clear, achievable goals with regular check-ins"""

        if 'attendance' in message_lower:
            return """Attendance is critical for academic success. To improve attendance:
- Identify and address barriers (transportation, health, family issues)
- Implement positive reinforcement for good attendance
- Communicate regularly with parents
- Make learning engaging and relevant
- Provide support for students facing challenges"""

        if 'model' in message_lower or 'prediction' in message_lower:
            return """The prediction system uses machine learning models (Random Forest and SVM) trained on historical student data.
The models analyze factors like marks, attendance, assignments, and participation to predict which students may need support.
The system provides probability scores and explanations to help educators make informed decisions."""

        return """I can help you understand student performance predictions and suggest interventions.

You can ask me:
- \"Why is student X at risk?\"
- \"How can I help improve this student's performance?\"
- \"What interventions work for at-risk students?\"
- \"Explain the prediction model\"
- \"What factors indicate a student is at risk?\"

Please provide more details about your question, and I'll do my best to assist you."""

    def _explain_risk(self, student_context: Dict) -> str:
        if not student_context:
            return "Unable to retrieve student information."
        risk_factors = student_context.get('risk_factors', [])
        if not risk_factors:
            return "This student shows positive indicators and is not currently classified as at-risk."
        explanation = f"Student {student_context.get('student_id')} is at risk due to:\n\n"
        for i, factor in enumerate(risk_factors, 1):
            explanation += f"{i}. {factor}\n"
        explanation += "\nThese factors suggest the student may benefit from additional support and intervention."
        return explanation

    def _suggest_improvements(self, student_context: Dict) -> str:
        if not student_context:
            return "Unable to retrieve student information."
        risk_factors = student_context.get('risk_factors', [])
        strategies = self.rag_pipeline.retrieve_intervention_strategies(risk_factors)
        response = f"To help Student {student_context.get('student_id')} improve, consider:\n\n"
        for i, strategy in enumerate(strategies[:5], 1):
            response += f"{i}. {strategy}\n"
        response += "\nFocus on building a supportive relationship and addressing underlying issues."
        return response

    def _suggest_interventions(self, student_context: Dict) -> str:
        if not student_context:
            return "Unable to retrieve student information."
        risk_factors = student_context.get('risk_factors', [])
        strategies = self.rag_pipeline.retrieve_intervention_strategies(risk_factors)
        response = f"Intervention Plan for Student {student_context.get('student_id')}:\n\n"
        response += "IMMEDIATE ACTIONS:\n"
        for strategy in strategies[:3]:
            response += f"• {strategy}\n"
        response += "\nSHORT-TERM GOALS (1-2 months):\n"
        response += "• Improve attendance to 85%+\n"
        response += "• Complete all assignments on time\n"
        response += "• Show measurable academic improvement\n"
        response += "\nLONG-TERM STRATEGIES:\n"
        for strategy in strategies[3:6]:
            response += f"• {strategy}\n"
        response += "\nMonitor progress weekly and adjust the plan as needed."
        return response

    def get_conversation_history(self) -> List[Dict]:
        return self.conversation_history

    def clear_history(self):
        self.conversation_history = []


def get_chatbot_response(message: str, student_id: Optional[str] = None, students_data: Optional[List[Dict]] = None) -> str:
    chatbot = StudentPerformanceChatbot()
    if students_data:
        chatbot.index_student_data(students_data)
    return chatbot.chat(message, student_id)

File: frontend/src/components/Chatbot.js
import React, { useState, useEffect, useRef } from 'react';
import './Chatbot.css';

const Chatbot = ({ predictionResults }) => {
  const [isOpen, setIsOpen] = useState(false);
  const [chatMessage, setChatMessage] = useState('');
  const [chatHistory, setChatHistory] = useState([]);
  const [isTyping, setIsTyping] = useState(false);
  const chatEndRef = useRef(null);

  useEffect(() => {
    chatEndRef.current?.scrollIntoView({ behavior: 'smooth' });
  }, [chatHistory]);

  const handleChatSubmit = async (e) => {
    e.preventDefault();
    if (!chatMessage.trim()) return;

    const userMessage = { role: 'user', content: chatMessage, timestamp: new Date() };
    setChatHistory([...chatHistory, userMessage]);
    setChatMessage('');
    setIsTyping(true);

    try {
      const response = await fetch('http://localhost:5000/api/chatbot', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          message: chatMessage,
          students_data: predictionResults?.predictions || [],
          context: {
            total_students: predictionResults?.total_students || 0,
            at_risk_count: predictionResults?.at_risk_count || 0,
            at_risk_percentage: predictionResults?.at_risk_percentage || 0
          }
        }),
      });

      const data = await response.json();

      const botResponse = {
        role: 'assistant',
        content: data.response || data.message || 'I apologize, but I encountered an issue. Please try again.',
        timestamp: new Date()
      };

      setTimeout(() => {
        setChatHistory(prev => [...prev, botResponse]);
        setIsTyping(false);
      }, 500);

    } catch (_error) {
      const botResponse = {
        role: 'assistant',
        content: generateFallbackResponse(chatMessage),
        timestamp: new Date()
      };

      setTimeout(() => {
        setChatHistory(prev => [...prev, botResponse]);
        setIsTyping(false);
      }, 500);
    }
  };

  const generateFallbackResponse = (message) => {
    const lowerMessage = message.toLowerCase();

    if (lowerMessage.includes('student') && (lowerMessage.match(/\d+/) || lowerMessage.includes('specific'))) {
      return `To analyze a specific student, please provide their ID. Based on the current predictions, ${predictionResults?.at_risk_count || 0} students are at risk. Click on individual students in the results table for detailed analysis.`;
    }

    if (lowerMessage.includes('at risk') || lowerMessage.includes('failing')) {
      const atRiskCount = predictionResults?.at_risk_count || 0;
      const atRiskPct = predictionResults?.at_risk_percentage || 0;
      return `Currently, ${atRiskCount} students (${atRiskPct.toFixed(1)}%) are identified as at-risk. These students show indicators like low marks, poor attendance, or declining performance. Early intervention can significantly improve their outcomes.`;
    }

    if (lowerMessage.includes('intervention') || lowerMessage.includes('help') || lowerMessage.includes('improve')) {
      return `Effective interventions include:\n\n1. **Personalized Support**: One-on-one tutoring tailored to individual needs\n2. **Regular Monitoring**: Weekly check-ins to track progress\n3. **Parent Engagement**: Involve parents in the improvement plan\n4. **Peer Mentoring**: Pair struggling students with high-achievers\n5. **Study Skills Training**: Teach time management and learning strategies\n6. **Address Root Causes**: Identify and resolve attendance or personal issues\n\nEarly intervention is key to success!`;
    }

    if (lowerMessage.includes('attendance')) {
      return `Attendance is one of the strongest predictors of academic success. To improve attendance:\n\n• Identify barriers (transportation, health, family issues)\n• Implement positive reinforcement programs\n• Communicate regularly with parents\n• Make learning engaging and relevant\n• Provide support for students facing challenges\n\nStudents with <75% attendance are at significantly higher risk.`;
    }

    if (lowerMessage.includes('mark') || lowerMessage.includes('grade') || lowerMessage.includes('score')) {
      return `Academic performance is analyzed across multiple subjects. Students with average marks below 50% are flagged as at-risk. Key factors affecting marks include:\n\n• Understanding of core concepts\n• Study habits and time management\n• Attendance and class participation\n• Assignment completion\n• Previous performance trends\n\nConsider subject-specific tutoring for students struggling in particular areas.`;
    }

    if (lowerMessage.includes('model') || lowerMessage.includes('prediction') || lowerMessage.includes('accurate') || lowerMessage.includes('how does')) {
      return `The system uses machine learning models (Random Forest and SVM) trained on historical student data. The models analyze:\n\n• Academic marks across subjects\n• Attendance rates\n• Assignment completion\n• Performance trends\n• Class participation\n\nTypical accuracy: 80-90%\n\nPredictions provide probability scores to help educators identify students who may need support. They should be used as one tool among many for decision-making.`;
    }

    if (lowerMessage.includes('factor') || lowerMessage.includes('indicator') || lowerMessage.includes('cause')) {
      return `Key risk factors identified by the model:\n\n🔴 **High Risk Indicators:**\n• Average marks < 40%\n• Attendance < 60%\n• Declining performance trend\n• Low assignment completion (<50%)\n\n🟡 **Moderate Risk Indicators:**\n• Marks between 40-50%\n• Attendance 60-75%\n• Inconsistent performance\n\n🟢 **Positive Indicators:**\n• Marks > 70%\n• Attendance > 85%\n• Improving trends\n• High engagement`;
    }

    if (lowerMessage.includes('recommend') || lowerMessage.includes('suggest') || lowerMessage.includes('strategy')) {
      const atRiskPct = predictionResults?.at_risk_percentage || 0;
      
      if (atRiskPct > 50) {
        return `🚨 **CRITICAL SITUATION**: Over 50% of students at risk.\n\n**Immediate Actions:**\n1. Emergency staff meeting to review situation\n2. Class-wide diagnostic assessment\n3. Implement intensive support program\n4. Review teaching methods and curriculum\n5. Engage parents with urgent communication\n6. Consider additional resources and tutoring\n\nThis requires systemic intervention, not just individual support.`;
      } else if (atRiskPct > 30) {
        return `⚠️ **HIGH ALERT**: Significant portion at risk.\n\n**Recommended Actions:**\n1. Form student support team\n2. Develop targeted intervention plans\n3. Increase parent-teacher communication\n4. Implement peer tutoring program\n5. Schedule regular progress reviews\n6. Provide additional learning resources\n\nFocus on both individual and small-group interventions.`;
      } else {
        return `✅ **MANAGEABLE SITUATION**: Focus on individual students.\n\n**Recommended Actions:**\n1. Schedule one-on-one meetings with at-risk students\n2. Create personalized learning plans\n3. Monitor progress weekly\n4. Maintain open communication with parents\n5. Celebrate small wins to build confidence\n6. Connect students with appropriate resources\n\nEarly identification gives you time for effective intervention!`;
      }
    }

    if (lowerMessage.includes('class') || lowerMessage.includes('overall') || lowerMessage.includes('summary')) {
      const total = predictionResults?.total_students || 0;
      const atRisk = predictionResults?.at_risk_count || 0;
      const safe = total - atRisk;
      return `📊 **Class Overview:**\n\n• Total Students: ${total}\n• At Risk: ${atRisk} students\n• Not at Risk: ${safe} students\n• Risk Percentage: ${predictionResults?.at_risk_percentage?.toFixed(1) || 0}%\n\nThe predictions help you identify which students need immediate attention. Focus your resources on the at-risk group while maintaining support for others.`;
    }

    if (lowerMessage.includes('how to use') || lowerMessage.includes('how do i')) {
      return `📖 **How to Use This System:**\n\n1. **Review Results**: Check the predictions table for at-risk students\n2. **Detailed Analysis**: Click on individual students to see risk factors\n3. **Action Items**: Note the recommendations for each student\n4. **Track Progress**: Re-upload data periodically to monitor improvements\n5. **Ask Questions**: Use this chatbot to get insights and advice\n\nRemember: Predictions are tools to guide decisions, not replace professional judgment.`;
    }

    return `I'm here to help you understand student performance predictions and provide guidance on interventions.\n\n**You can ask me about:**\n• Why specific students are at risk\n• What interventions work best\n• How to improve attendance or marks\n• Understanding the prediction model\n• Class-level statistics and recommendations\n• Specific risk factors and indicators\n\n**Try asking:**\n• "Why are students at risk?"\n• "What interventions do you recommend?"\n• "How can I improve attendance?"\n• "Tell me about the at-risk students"\n• "What factors indicate risk?"\n\nWhat would you like to know?`;
  };

  const handleSuggestedQuestion = (question) => {
    setChatMessage(question);
  };

  const toggleChat = () => {
    setIsOpen(!isOpen);
    if (!isOpen && chatHistory.length === 0) {
      const welcomeMessage = {
        role: 'assistant',
        content: `👋 Hi! I'm your AI teaching assistant. I can help you understand the student performance predictions and suggest interventions.\n\nWhat would you like to know?`,
        timestamp: new Date()
      };
      setChatHistory([welcomeMessage]);
    }
  };

  const clearChat = () => {
    setChatHistory([]);
    const welcomeMessage = {
      role: 'assistant',
      content: `Chat cleared! How can I help you?`,
      timestamp: new Date()
    };
    setChatHistory([welcomeMessage]);
  };

  return (
    <div className={`chatbot-widget ${isOpen ? 'open' : 'closed'}`}>
      {!isOpen && (
        <button className="chatbot-toggle-button" onClick={toggleChat}>
          <span className="chat-icon">💬</span>
          <span className="chat-label">AI Assistant</span>
        </button>
      )}

      {isOpen && (
        <div className="chatbot-window">
          <div className="chatbot-header">
            <div className="chatbot-header-content">
              <span className="chatbot-icon">🤖</span>
              <div className="chatbot-title">
                <h3>AI Teaching Assistant</h3>
                <span className="chatbot-status">● Online</span>
              </div>
            </div>
            <div className="chatbot-actions">
              <button className="chatbot-action-btn" onClick={clearChat} title="Clear chat">
                🗑️
              </button>
              <button className="chatbot-action-btn" onClick={toggleChat} title="Close">
                ✕
              </button>
            </div>
          </div>

          <div className="chatbot-messages">
            {chatHistory.length === 0 ? (
              <div className="chat-welcome">
                <div className="welcome-icon">👋</div>
                <h3>How can I assist you?</h3>
                <p>Ask me about student predictions, interventions, or teaching strategies.</p>
                
                <div className="suggested-questions">
                  <button onClick={() => handleSuggestedQuestion('Why are students at risk?')}>
                    Why are students at risk?
                  </button>
                  <button onClick={() => handleSuggestedQuestion('What interventions do you recommend?')}>
                    Recommended interventions?
                  </button>
                  <button onClick={() => handleSuggestedQuestion('Tell me about the at-risk students')}>
                    At-risk student details
                  </button>
                  <button onClick={() => handleSuggestedQuestion('How can I improve class performance?')}>
                    Improve performance
                  </button>
                </div>
              </div>
            ) : (
              <>
                {chatHistory.map((msg, idx) => (
                  <div key={idx} className={`chat-message ${msg.role}`}>
                    <div className="message-avatar">
                      {msg.role === 'user' ? '👤' : '🤖'}
                    </div>
                    <div className="message-bubble">
                      <div className="message-content">
                        {msg.content.split('\n').map((line, i) => (
                          <React.Fragment key={i}>
                            {line}
                            {i < msg.content.split('\n').length - 1 && <br />}
                          </React.Fragment>
                        ))}
                      </div>
                      <div className="message-time">
                        {msg.timestamp?.toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' })}
                      </div>
                    </div>
                  </div>
                ))}
                {isTyping && (
                  <div className="chat-message assistant">
                    <div className="message-avatar">🤖</div>
                    <div className="message-bubble typing">
                      <div className="typing-indicator">
                        <span></span>
                        <span></span>
                        <span></span>
                      </div>
                    </div>
                  </div>
                )}
                <div ref={chatEndRef} />
              </>
            )}
          </div>

          <form onSubmit={handleChatSubmit} className="chatbot-input-form">
            <input
              type="text"
              value={chatMessage}
              onChange={(e) => setChatMessage(e.target.value)}
              placeholder="Type your question here..."
              className="chatbot-input"
              disabled={isTyping}
            />
            <button 
              type="submit" 
              className="chatbot-send-button"
              disabled={!chatMessage.trim() || isTyping}
            >
              ➤
            </button>
          </form>
        </div>
      )}
    </div>
  );
};

export default Chatbot;
